File: api-app/src/main/java/de/muspellheim/activitysampling/api/domain/WorkingDay.java
/* Copyright (c) 2025 Falko Schumann. All rights reserved. MIT license. */

package de.muspellheim.activitysampling.api.domain;

import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import lombok.NonNull;

public record WorkingDay(@NonNull LocalDate date, @NonNull List<Activity> activities) {

  public static List<WorkingDay> from(List<Activity> recentActivities) {
    var workingDays = new ArrayList<WorkingDay>();
    LocalDate date = null;
    var activities = new ArrayList<Activity>();
    for (var activity : recentActivities) {
      // TODO Handle time zone
      var activityDate = activity.timestamp().atZone(ZoneId.systemDefault()).toLocalDate();
      if (!activityDate.equals(date)) {
        if (date != null) {
          workingDays.add(new WorkingDay(date, List.copyOf(activities)));
        }

        date = activityDate;
        activities.clear();
      }
      activities.add(activity);
    }
    if (date != null) {
      workingDays.add(new WorkingDay(date, List.copyOf(activities)));
    }
    return workingDays;
  }
}

File: src/shared/domain/working_day.ts
// Copyright (c) 2026 Falko Schumann. All rights reserved. MIT license.

import { RecentActivity } from "./recent_activity";

export class WorkingDay {
  static compare(a: WorkingDay, b: WorkingDay) {
    return Temporal.PlainDate.compare(a.date, b.date);
  }

  static create({
    date,
    activities = [],
  }: {
    date: Temporal.PlainDateLike;
    activities?: RecentActivity[];
  }) {
    return new WorkingDay(date, activities);
  }

  static createTestInstance({
    date = "2026-03-29",
    activities = [RecentActivity.createTestInstance()],
  }: {
    date?: Temporal.PlainDateLike;
    activities?: RecentActivity[];
  } = {}) {
    return WorkingDay.create({
      date,
      activities,
    });
  }

  readonly date;
  readonly activities;

  private constructor(
    date: Temporal.PlainDateLike,
    activities: RecentActivity[],
  ) {
    this.date = Temporal.PlainDate.from(date);
    this.activities = activities.map((activity) =>
      RecentActivity.create(activity),
    );
  }
}
